src/models/filter.rs

```diff
@@ -193,68 +208,73 @@ impl SourceCodeUnit {
     &self, filter: Filter, node: Node, rule_store: &mut RuleStore,
     substitutions: &HashMap<String, String>,
   ) -> bool {
-    // This ensures that the below while loop considers the current node too when checking for filters.
-    // It does not make sense to check for filter if current node is a "leaf" node.
-    let mut initial_node = node;
+    let mut node_to_check = node;
+    let instantiated_filter = filter.instantiate(substitutions);
 
-    // No enclosing node is provided
-    if filter.enclosing_node().get_query().as_str() == DEFAULT_ENCLOSING_QUERY {
-      // Get the enclosing node matching the pattern specified in the filter (`filter.enclosing_node`)
-      self._check_current_node(filter, rule_store, substitutions, initial_node)
-    } else {
+    // If there is a not_enclosing_node is provided
+    let query = instantiated_filter.not_enclosing_node();
+    if !query.get_query().is_empty() {
+      // No ancestor should match with it
+      if self
+        ._get_match_from_ancestor(rule_store, node_to_check, query)
+        .is_some()
+      {
+        return false;
+      }
+    }
+    // If an enclosing node is provided
+    let query = instantiated_filter.enclosing_node();
+    if !query.get_query().is_empty() {
+      // This ensures that the below while loop considers the current node too when checking for filters.
```

Comment:
update comments. There is no while loop below now. 
also it could be worth moving this logic within `_get_match_from_ancestor` if possible. (Assuming it is not used elsewhere)